src/document.js

```diff
@@ -500,6 +523,57 @@ const documentsMain = {
 				$(document.body).removeClass('claro')
 			})
 		},
+
+		initiateOpenLocally() {
+			OC.dialogs.confirmDestructive(
+				t('richdocuments', 'When opening a file locally, the document will close for all users currently viewing the document.'),
+				t('richdocuments', 'Open file locally'),
+				{
+					type: OC.dialogs.YES_NO_BUTTONS,
+					confirm: t('richdocuments', 'Open locally'),
+					confirmClasses: 'error',
+					cancel: t('richdocuments', 'Continue editing online'),
+				},
+				(decision) => {
+					if (!decision) {
+						return
+					}
+					documentsMain.openingLocally = true
+					PostMessages.sendWOPIPostMessage('loolframe', 'Get_Views')
+				}
+			)
+		},
+
+		removeViews(views) {
+			PostMessages.sendWOPIPostMessage('loolframe', 'Action_Save', {
+				DontTerminateEdit: false,
+				DontSaveIfUnmodified: false,
+				Notify: false,
+			})
+
+			views.forEach((view) => {
+				PostMessages.sendWOPIPostMessage('loolframe', 'Action_RemoveView', { ViewId: Number(view.ViewId) })
+			})
+		},
+	},
+
+	unlockFile() {
+		const unlockUrl = getRootUrl() + '/index.php/apps/richdocuments/wopi/files/' + documentsMain.fileId
+		const unlockConfig = {
+			headers: { 'X-WOPI-Override': 'UNLOCK' }
+		}
+		axios.post(unlockUrl, { access_token: documentsMain.token }, unlockConfig)
+			.catch(_ => {}) // Unlocking failed, possibly because file was not locked, we want to proceed regardless.
+			.then(() => {
+				documentsMain.openLocally()
+			})
+	},
+
+	openLocally() {
+		if (documentsMain.openingLocally) {
+			documentsMain.openingLocally = false
+			window.location.href = 'nc://open/' + Config.get('userId') + '@' + window.location.host + documentsMain.fullPath
```

Comment:
Maybe we can adapt the logic from https://github.com/nextcloud/server/pull/34171/files#diff-dedceb236d4709ee562b67512f8ad10bce9dc7a494e9b1cfdabe8ea8ad33ca25R2810-R2818 to also take the port into account